pkg/filter/headerlookup/headerlookup.go

```diff
@@ -0,0 +1,188 @@
+	}
+
+	// Spec defines header key and etcd prefix that form etcd key like /custom-data/{etcdPrefix}/{headerKey's value}.
+	// This /custom-data/{etcdPrefix}/{headerKey's value} is retrieved from etcd and HeaderSetters extract keys from the
+	// from the retrieved etcd item.
+	Spec struct {
+		HeaderKey     string              `yaml:"headerKey" jsonschema:"required"`
+		EtcdPrefix    string              `yaml:"etcdPrefix" jsonschema:"required"`
+		HeaderSetters []*HeaderSetterSpec `yaml:"headerSetters" jsonschema:"required"`
+	}
+)
+
+// Validate validates spec.
+func (spec Spec) Validate() error {
+	if spec.HeaderKey == "" {
+		return fmt.Errorf("headerKey is required")
+	}
+	if spec.EtcdPrefix == "" {
+		return fmt.Errorf("etcdPrefix is required")
+	}
+	if len(spec.HeaderSetters) < 1 {
+		return fmt.Errorf("at least one headerSetter is required")
+	}
+	for _, hs := range spec.HeaderSetters {
+		if hs.EtcdKey == "" {
+			return fmt.Errorf("headerSetters[i].etcdKey is required")
+		}
+		if hs.HeaderKey == "" {
+			return fmt.Errorf("headerSetters[i].headerKey is required")
+		}
+	}
+	return nil
+}
+
+// Kind returns the kind of HeaderLookup.
+func (hl *HeaderLookup) Kind() string {
+	return Kind
+}
+
+// DefaultSpec returns the default spec of HeaderLookup.
+func (hl *HeaderLookup) DefaultSpec() interface{} {
+	return &Spec{}
+}
+
+// Description returns the description of HeaderLookup.
+func (hl *HeaderLookup) Description() string {
+	return "HeaderLookup enriches request headers per request, looking up values from etcd."
+}
+
+// Results returns the results of HeaderLookup.
+func (hl *HeaderLookup) Results() []string {
+	return results
+}
+
+// Init initializes HeaderLookup.
+func (hl *HeaderLookup) Init(filterSpec *httppipeline.FilterSpec) {
+	hl.filterSpec, hl.spec = filterSpec, filterSpec.FilterSpec().(*Spec)
+	if filterSpec.Super() != nil && filterSpec.Super().Cluster() != nil {
+		hl.cluster = filterSpec.Super().Cluster()
+	}
+}
+
+// Inherit inherits previous generation of HeaderLookup.
+func (hl *HeaderLookup) Inherit(filterSpec *httppipeline.FilterSpec, previousGeneration httppipeline.Filter) {
+	previousGeneration.Close()
+	hl.Init(filterSpec)
+}
+
+func parseYamlCreds(entry string) (map[string]string, error) {
+	var err error
+	defer func() {
+		if err := recover(); err != nil {
+			err = fmt.Errorf("could not marshal custom-data, ensure that it's valid yaml")
+		}
+	}()
+	yamlMap := make(map[string]string)
+	yamltool.Unmarshal([]byte(entry), &yamlMap)
+	return yamlMap, err
+}
+
+func (hl *HeaderLookup) lookup(headerVal string) (map[string]string, error) {
+	etcdKey := customDataPrefix + strings.TrimPrefix(hl.spec.EtcdPrefix, "/") + headerVal
+	etcdVal, err := hl.cluster.Get(etcdKey)
+	if err != nil {
+		return nil, err
+	}
+	if etcdVal == nil {
+		return nil, fmt.Errorf("no data found")
+	}
+	result := make(map[string]string, len(hl.spec.HeaderSetters))
+	etcdValues, err := parseYamlCreds(*etcdVal)
```

Comment:
That's a good point! I think LRU cache and [Syncer](https://github.com/megaease/easegress/blob/main/pkg/cluster/syncer.go) to invalidate the cache items would help.

Reply:
https://github.com/megaease/easegress/pull/454/commits/1565bab1ec2fab5058d692e19e764089d65ba737 adds LRU cache around `hl.cluster.Get` to avoid making this filter bottleneck. A separate goroutine listens Syncer for updates and removes modified or missing keys.